tokio-net/src/driver/reactor.rs

```diff
@@ -349,17 +346,15 @@ impl fmt::Debug for Reactor {
 // ===== impl Handle =====
 
 impl Handle {
-    #[doc(hidden)]
-    #[deprecated(note = "semantics were sometimes surprising, use Handle::default()")]
-    pub fn current() -> Handle {
-        // TODO: Should this panic on error?
-        HandlePriv::try_current()
-            .map(|handle| Handle {
-                inner: Some(handle),
-            })
-            .unwrap_or(Handle {
-                inner: Some(HandlePriv { inner: Weak::new() }),
-            })
+    /// Get a handle to the current reactor.
+    ///
+    /// # Returns
+    ///
+    /// - `Ok<Handle>` if there is reactor set for the current thread
+    /// - `Err` if there is no reactor set for the current thread
+    pub fn current() -> io::Result<Handle> {
```

Comment:
right now it always returns Ok, so what's point of retaining Result here?

I'd suggest to retain Default and point user to use it, rather than current()

But probably would be good to panic in debug mode, when we're not in runtime actually

Reply:
So the point of `Result` is it does not always return `Ok` as shown in a test case [here](https://github.com/kleimkuhler/tokio/blob/54dc256906c56afaa11f1454318bf705a8be3817/tokio-net/tests/bind_resource.rs#L15-L16).

If a user attempts to make a resource outside of a runtime, then `CURRENT_REACTOR` is not set on that thread and this function returns an `Err`.

I think the ask is probably is this an incorrect use of the API? If so, maybe this should just panic here?

I had concern about this being part of the `Default` impl because I don't think `default` should be fallible, and that will be possible if the thread doesn't have the variable set properly.

Reply:
`current` now returns a handle to the current reactor, otherwise it panics if there is not one set. This is only called when registering an I/O resource with a reactor, so it is expected that this is only called from within a runtime.